### PR TITLE
Localize link text audit

### PR DESCRIPTION
**Pull request by raffaelj**
**Summary**

The link-text audit produces false positives, because it is not localized. I changed the anchor-elements gatherer and the link-text audit to check for the current language.

**Related Issues/PRs**

See discussion in https://github.com/GoogleChrome/lighthouse/issues/14845

### CODE REVIEW

File: core/gather/gatherers/anchor-elements.js [hunk 1]
```diff
@@ -38,10 +38,71 @@ function collectAnchorElements() {
...
+        }
+
+        const elLang = el.getAttribute('lang');
+        const childrenWithLang = el.querySelectorAll('[lang]');
+
+        if (!childrenWithLang.length) {
+          return elLang || currentLang || '';
+        } else {
+          // recursive call
+          return getLangOfInnerText(el, elLang || currentLang || '');
```
Comment (adamraine) [inline core/gather/gatherers/anchor-elements.js:82]:
```suggestion
          return getLangOfInnerText(el, elLang || currentLang || '');
```

File: cli/test/fixtures/seo/seo-failure-cases.html [hunk 2]
```diff
@@ -5,7 +5,7 @@
  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 -->
 
-<html>
+<html lang="en">
```
Comment (connorjclark) [inline cli/test/fixtures/seo/seo-failure-cases.html:8]: It'd be good to update this smoke test to include cases that exercise the new logic (`getLangOfInnerText`). The expectations for this smoke tests are found in `seo-failing.js`. If the current smoke expectations were more explicit,  it would have this for the `link-text` audit (currently it just checks for 4 failures, but it'd be good to be more explicit now):

```json
'link-text': {
  score: 0,
  displayValue: '4 links found',
  details: {
    items: [
      {text: 'click this'},
      {text: 'click this'},
      {text: 'CLICK THIS'},
      {text: 'click this'},
    ],
  },
}
```

Can you devise some cases here to make sure `getLangOfInnerText` works as expected? If you need to create a new file HTML in `fixtures/seo` and a new accompanying expectation file in `test-definitions/seo-mixed-language.js`, that's OK.

File: core/gather/gatherers/anchor-elements.js [hunk 3]
```diff
@@ -38,10 +38,71 @@ function collectAnchorElements() {
...
 
+  // Check, if document has only one lang attribute in opening html tag. If so,
+  // there is no need to run the `getLangOfInnerText()` function with multiple
+  // possible DOM traversals
+  /** @type {Array<HTMLElement|SVGElement>} */
+  // @ts-expect-error - put into scope via stringification
+  const langElements = getElementsInDocument('[lang]'); // eslint-disable-line no-undef
+  const docHasOnlyHtmlLangAttr = (langElements.length === 1) &&
+    langElements[0].nodeName === 'HTML';
+  const lang = !docHasOnlyHtmlLangAttr ? null : langElements[0].getAttribute('lang');
```
Comment (connorjclark) [inline core/gather/gatherers/anchor-elements.js:104]: this needs to be normalized some how. Consider the valid language code `en-US`.

I don't recall how we could actually normalize to the base language code. In `shared/localization/locales.js` we form a mapping explicitly. I suppose here we could just split by `-` and return the first part to use as the loclae.

File: core/gather/gatherers/anchor-elements.js [hunk 4]
```diff
@@ -93,6 +93,16 @@ function collectAnchorElements() {
   // @ts-expect-error - put into scope via stringification
   const anchorElements = getElementsInDocument('a'); // eslint-disable-line no-undef
 
+  // Check, if document has only one lang attribute in opening html tag. If so,
+  // there is no need to run the `getLangOfInnerText()` function with multiple
+  // possible DOM traversals
+  /** @type {Array<HTMLElement|SVGElement>} */
+  // @ts-expect-error - put into scope via stringification
+  const langElements = getElementsInDocument('[lang]'); // eslint-disable-line no-undef
```
Comment (connorjclark) [inline core/gather/gatherers/anchor-elements.js:101]: Could this be more clear by instead explicitly checking `document.documentElement.lang`, and searching for further codes at `body [lang]`?

**Comment by raffaelj** [comment]
@connorjclark @adamraine Thanks for the review. The easy ones are fixed. I wrote a smoke test, but I don't have more time today. I'll do some cleanup and push the changes tomorrow.

I'll have a look at the other suggestions this week (or weekend).

File: cli/test/fixtures/seo/seo-tester.html [hunk 5]
```diff
@@ -5,7 +5,7 @@
  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 -->
 
-<html>
+<html lang="en">
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply cli/test/fixtures/seo/seo-tester.html:8]: We discussed this, and we think it'd be best to do the current behavior of the audit when the document does not declare a language. I think that's what the PR is currently doing, but some tests in the audit would be good to verify.

File: core/test/audits/seo/link-text-test.js [hunk 6]
```diff
@@ -10,15 +10,15 @@ import LinkTextAudit from '../../../audits/seo/link-text.js';
 
 describe('SEO: link text audit', () => {
   it('fails when link with non descriptive text is found', () => {
-    const invalidLink = {href: 'https://example.com/otherpage.html', text: 'click here', rel: ''};
```
Comment (connorjclark) [inline core/test/audits/seo/link-text-test.js:13]: I think adding a new test or two that has non-english text that is bad would be clearer. All these existing test cases could leave off `lang: 'en'` since it doesn't modify the thing they are asserting; plus that gives us coverage for the common case of there being no document language.

File: cli/test/fixtures/seo/seo-failure-cases.html (hunk 2 above)
Comment (raffaelj) [reply cli/test/fixtures/seo/seo-failure-cases.html:8]: I changed the smoke test `seo-failing` to be more explicit, including the language attribute and a German example. I also added a new smoke test `seo-mixed-language` with many edge cases (German and English examples).

File: cli/test/fixtures/seo/seo-tester.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (raffaelj) [reply cli/test/fixtures/seo/seo-tester.html:8]: Just to be sure - What do you mean with "current behavior"? The current behavior before this PR or the current behavior of this PR?

My implementation ignores links if the language is not declared. I also added a new test case to `core/test/audits/seo/link-text-test.js` to reflect that new behavior.

File: core/audits/seo/link-text.js [hunk 7]
```diff
@@ -126,7 +151,21 @@ class LinkText extends Audit {
...
+
+        if (searchTerm) {
+          // walk through lang code, e. g. "en-US" or "zh-cmn-Hans-CN"
+          let lang = link.textLang;
+          while (lang) {
+            if (nonDescriptiveLinkTexts[lang] &&
+              nonDescriptiveLinkTexts[lang].has(searchTerm)) {
+              return true;
+            }
+            lang = lang.substring(0, lang.lastIndexOf('-'));
```
[the remark this answers is not in the document]
Comment (raffaelj) [reply core/audits/seo/link-text.js:164]: I'm not that deep into linguistics. But valid language codes like `zh-cmn-Hans-CN` exist and there might be differences between e. g. Cantonese and Simplified Chinese. So instead of speculating and to keep it future proof (support more than one segment if needed), I decided to use the loop.

File: core/test/audits/seo/link-text-test.js (hunk 6 above)
Comment (raffaelj) [reply core/test/audits/seo/link-text-test.js:13]: I cleaned it up and added more tests.

File: core/gather/gatherers/anchor-elements.js (hunk 3 above)
Comment (raffaelj) [reply core/gather/gatherers/anchor-elements.js]: I wouldn't normalize here. The gatherer just collects the lang code and passes it to the audit. If it should get normalized, then the last tool in the chain (the audit) should do that.

See also my response below to @adamraine about keeping the loop over language code segments.

**Comment by raffaelj** [comment]
I think, I addressed all annotations.

If you really want to change the audit to use only the first segment of a language code (e. g. `en` from `en-US`), I could do it. But I still think, it's better to keep my current implementation.

A few questions are still open from https://github.com/GoogleChrome/lighthouse/issues/14845#issuecomment-1483264251:

* headline "Fallbacks"
  * fallback to pragma-set-default-language or HTTP header
* headline "Other thoughts"
  * more exceptions
  * change weight of audit

**Review by ckopoctb999 — COMMENTED** [review]
[s[moke](https://github.com/GoogleChrome/lighthouse/actions/workflows/smoke.yml)](raffaelj:localize-link-text-audit)

**Comment by raffaelj** [comment]
@connorjclark @adamraine Can I do anything to finish this PR?

**Review by adamraine — COMMENTED** [review]
Sorry for the delay @raffaelj, and thanks for being patient

**Comment by raffaelj** [comment]
> Sorry for the delay, and thanks for being patient

Now I have to add another delay. I see 10 changed files from 3-6 months ago and I have to read them all again before I can answer. I'm not sure, if I find some time during the next weeks. If someone else wants to take over to speed things up - just go ahead.

**Comment by connorjclark** [comment]
I updated this PR over in #16526 - thank you @raffaelj for the time you spent setting this up!
